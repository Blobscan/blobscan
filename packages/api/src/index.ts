import type { inferRouterInputs, inferRouterOutputs } from "@trpc/server";

import type { AppRouter } from "./app-router";

export { appRouter, type AppRouter } from "./app-router";
export { createTRPCContext, createTRPCInnerContext } from "./context";
export type {
  TRPCInnerContext as TRCInnerContext,
  TRPCContext,
} from "./context";

export * from "@trpc/server/adapters/express";

export { env } from "./env";
export { gracefulShutdown } from "./graceful-shutdown";
export { metricsHandler } from "./instrumentation";
export type {
  BlobStorage,
  Rollup,
  Metric,
  MetricHistogram,
  MetricHistogramBucket,
  Metrics,
  MetricsClient,
} from "@blobscan/db";

/**
 * Inference helpers for input types
 * @example type HelloInput = RouterInputs['example']['hello']
 **/
export type RouterInputs = inferRouterInputs<AppRouter>;

/**
 * Inference helpers for output types
 * @example type HelloOutput = RouterOutputs['example']['hello']
 **/
export type RouterOutputs = inferRouterOutputs<AppRouter>;
