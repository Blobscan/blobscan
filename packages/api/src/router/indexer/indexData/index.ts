import type { Blob, Transaction, OmittableFields, Block } from "@blobscan/db";

import { tracer } from "../../../instrumentation";
import { jwtAuthedProcedure } from "../../../procedures";
import {
  calculateBlobGasPrice,
  calculateBlobSize,
  getEIP2028CalldataGas,
} from "../../../utils/ethereum";
import { INDEXER_PATH } from "../constants";
import type { IndexDataInput } from "./schema";
import { indexDataOutputSchema } from "./schema";
import { indexDataInputSchema } from "./schema";

function createDBTransactions({
  blobs,
  block,
  transactions,
}: IndexDataInput): Omit<Transaction, OmittableFields>[] {
  return transactions.map<Omit<Transaction, OmittableFields>>(
    ({ blockNumber, from, gasPrice, hash, maxFeePerBlobGas, to }) => {
      const txBlob: IndexDataInput["blobs"][0] | undefined = blobs.find(
        (b) => b.txHash === hash
      );

      if (!txBlob) {
        throw new Error(`Blob for transaction ${hash} not found`);
      }

      return {
        blockNumber,
        hash,
        fromId: from,
        toId: to,
        gasPrice,
        blobGasPrice: calculateBlobGasPrice(block.excessBlobGas),
        maxFeePerBlobGas,
        blobAsCalldataGasUsed: getEIP2028CalldataGas(txBlob.data),
      };
    }
  );
}

function createDBBlock(
  {
    block: { blobGasUsed, excessBlobGas, hash, number, slot, timestamp },
  }: IndexDataInput,
  dbTxs: { blobAsCalldataGasUsed: bigint }[]
): Omit<Block, OmittableFields> {
  const blobAsCalldataGasUsed = dbTxs.reduce(
    (acc, tx) => acc + tx.blobAsCalldataGasUsed,
    BigInt(0)
  );

  return {
    number,
    hash,
    timestamp: new Date(timestamp * 1000),
    slot,
    blobGasUsed,
    blobGasPrice: calculateBlobGasPrice(excessBlobGas),
    excessBlobGas,
    blobAsCalldataGasUsed,
  };
}

function createDBBlobs({
  blobs,
  block,
}: IndexDataInput): Omit<Blob, OmittableFields>[] {
  return blobs.map<Omit<Blob, OmittableFields>>((blob) => ({
    versionedHash: blob.versionedHash,
    commitment: blob.commitment,
    size: calculateBlobSize(blob.data),
    firstBlockNumber: block.number,
  }));
}

export const indexData = jwtAuthedProcedure
  .meta({
    openapi: {
      method: "PUT",
      path: `${INDEXER_PATH}/block-txs-blobs`,
      tags: ["indexer"],
      summary: "Index data in the database",
      protect: true,
    },
  })
  .input(indexDataInputSchema)
  .output(indexDataOutputSchema)
  .mutation(async ({ ctx: { prisma, blobStorageManager }, input }) => {
    const operations = [];

    // 1. Upload blobs' data to storages

    const newBlobs = await prisma.blob.filterNewBlobs(input.blobs);

    const blobUploadResults = await tracer.startActiveSpan(
      "blobs-upload",
      async (blobsUploadSpan) => {
        const result = (
          await Promise.all(
            newBlobs.map(async (b) =>
              blobStorageManager.storeBlob(b).then<{
                uploadRes: Awaited<
                  ReturnType<typeof blobStorageManager.storeBlob>
                >;
                blob: Awaited<
                  ReturnType<typeof prisma.blob.filterNewBlobs>
                >[number];
              }>((uploadRes) => ({
                blob: b,
                uploadRes,
              }))
            )
          )
        ).flat();

        blobsUploadSpan.end();

        return result;
      }
    );

    blobUploadResults.forEach(({ uploadRes: { errors }, blob }) => {
      if (errors.length) {
        const errorMsgs = errors.map((e) => `${e.storage}: ${e.error}`);

        console.warn(
          `Couldn't upload blob ${
            blob.versionedHash
          } to some of the storages: ${errorMsgs.join(", ")}`
        );
      }
    });

    const dbBlobStorageRefs = blobUploadResults.flatMap(
      ({ uploadRes: { references }, blob }) =>
        references.map((ref) => ({
          blobHash: blob.versionedHash,
          blobStorage: ref.storage,
          dataReference: ref.reference,
        }))
    );

    // 2. Prepare address, block, transaction and blob insertions

    // TODO: Create an upsert extension that set the `insertedAt` and the `updatedAt` field
    const now = new Date();

    const dbTxs = createDBTransactions(input);
    const dbBlock = createDBBlock(input, dbTxs);
    const dbBlobs = createDBBlobs(input);

    operations.push(
      prisma.block.upsert({
        where: { hash: input.block.hash },
        create: {
          ...dbBlock,
          insertedAt: now,
          updatedAt: now,
        },
        update: {
          ...dbBlock,
          updatedAt: now,
        },
      })
    );
    operations.push(
      prisma.address.upsertAddressesFromTransactions(input.transactions)
    );
    operations.push(prisma.transaction.upsertMany(dbTxs));
    operations.push(prisma.blob.upsertMany(dbBlobs));

    if (dbBlobStorageRefs.length) {
      operations.push(
        prisma.blobDataStorageReference.upsertMany(dbBlobStorageRefs)
      );
    }

    operations.push(
      prisma.blobsOnTransactions.createMany({
        data: input.blobs.map((blob) => ({
          blobHash: blob.versionedHash,
          txHash: blob.txHash,
          index: blob.index,
        })),
        skipDuplicates: true,
      })
    );

    // 3. Execute all database operations in a single transaction
    await prisma.$transaction(operations);
  });
