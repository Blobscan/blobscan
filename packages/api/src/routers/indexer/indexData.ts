import type { $Enums } from "@blobscan/db";

import { jwtAuthedProcedure } from "../../procedures";
import { INDEXER_PATH } from "./common";
import {
  indexDataInputSchema,
  indexDataOutputSchema,
} from "./indexData.schema";
import {
  createDBBlobs,
  createDBBlock,
  createDBTransactions,
} from "./indexData.utils";

type DBBlobStorageRef = {
  blobHash: string;
  blobStorage: $Enums.BlobStorage;
  dataReference: string;
};

export const indexData = jwtAuthedProcedure
  .meta({
    openapi: {
      method: "PUT",
      path: `${INDEXER_PATH}/block-txs-blobs`,
      tags: ["indexer"],
      summary: "indexes data in the database.",
      protect: true,
    },
  })
  .input(indexDataInputSchema)
  .output(indexDataOutputSchema)
  .mutation(
    async ({ ctx: { prisma, blobStorageManager, blobPropagator }, input }) => {
      const operations = [];

      const newBlobs = await prisma.blob.filterNewBlobs(input.blobs);

      let dbBlobStorageRefs: DBBlobStorageRef[] | undefined;

      // 2. Store blobs' data
      if (!blobPropagator) {
        const blobStorageOps = newBlobs.map(async (b) =>
          blobStorageManager.storeBlob(b).then((uploadRes) => ({
            blob: b,
            uploadRes,
          }))
        );
        const storageResults = (await Promise.all(blobStorageOps)).flat();

        dbBlobStorageRefs = storageResults.flatMap(
          ({ uploadRes: { references }, blob }) =>
            references.map((ref) => ({
              blobHash: blob.versionedHash,
              blobStorage: ref.storage,
              dataReference: ref.reference,
            }))
        );
      }

      // TODO: Create an upsert extension that set the `insertedAt` and the `updatedAt` field
      const now = new Date();

      // 3. Prepare address, block, transaction and blob insertions
      const dbTxs = createDBTransactions(input);
      const dbBlock = createDBBlock(input, dbTxs);
      const dbBlobs = createDBBlobs(input);

      operations.push(
        prisma.block.upsert({
          where: { hash: input.block.hash },
          create: {
            ...dbBlock,
            insertedAt: now,
            updatedAt: now,
          },
          update: {
            ...dbBlock,
            updatedAt: now,
          },
        })
      );
      operations.push(
        prisma.address.upsertAddressesFromTransactions(input.transactions)
      );
      operations.push(prisma.transaction.upsertMany(dbTxs));
      operations.push(prisma.blob.upsertMany(dbBlobs));

      if (dbBlobStorageRefs?.length) {
        operations.push(
          prisma.blobDataStorageReference.upsertMany(dbBlobStorageRefs)
        );
      }

      operations.push(
        prisma.blobsOnTransactions.createMany({
          data: input.blobs.map((blob) => ({
            blobHash: blob.versionedHash,
            txHash: blob.txHash,
            index: blob.index,
          })),
          skipDuplicates: true,
        })
      );

      // 4. Execute all database operations in a single transaction
      await prisma.$transaction(operations);

      // 5. Propagate blobs to storages
      if (blobPropagator) {
        await blobPropagator.propagateBlobs(newBlobs);
      }
    }
  );
