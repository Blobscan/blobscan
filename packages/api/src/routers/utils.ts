export const GAS_PER_BLOB = 131_072; // 2 ** 17
export const BLOB_GAS_LIMIT_PER_BLOCK = 786_432;

function isValidInt(number: number): boolean {
  const minInt = -2147_483_648;
  const maxInt = 2147_483_647;

  return number >= minInt && number <= maxInt;
}
/**
 *
 * Checks if the given string is a number
 */
export function isBlockNumber(number: string): boolean {
  const number_ = Number(number);

  return !isNaN(number_) && isValidInt(number_);
}
