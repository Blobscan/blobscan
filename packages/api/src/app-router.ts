import { z } from "@blobscan/zod";

import { publicProcedure } from "./procedures";
import { blobRouter } from "./routers/blob";
import { blobStoragesStateRouter } from "./routers/blob-storages-state";
import { blockRouter } from "./routers/block";
import { blockchainSyncStateRouter } from "./routers/blockchain-sync-state";
import { indexerRouter } from "./routers/indexer";
import { searchRouter } from "./routers/search";
import { statsRouter } from "./routers/stats";
import { transactionRouter } from "./routers/tx";
import { t } from "./trpc-client";

export const appRouter = t.router({
  block: blockRouter,
  tx: transactionRouter,
  blob: blobRouter,
  search: searchRouter,
  stats: statsRouter,
  indexer: indexerRouter,
  syncState: blockchainSyncStateRouter,
  blobStoragesState: blobStoragesStateRouter,
  healthcheck: publicProcedure
    .meta({
      openapi: {
        method: "GET",
        path: "/healthcheck",
        summary: "connection healthcheck.",
        tags: ["system"],
      },
    })
    .input(z.void())
    .output(z.string())
    .query(() => "yay!"),
});

// export type definition of API
export type AppRouter = typeof appRouter;
