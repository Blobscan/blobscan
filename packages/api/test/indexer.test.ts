import { TRPCError } from "@trpc/server";
import type { SpyInstance } from "vitest";
import {
  afterAll,
  afterEach,
  beforeAll,
  beforeEach,
  describe,
  expect,
  it,
  vi,
} from "vitest";

import type { Blob as PropagatorBlob } from "@blobscan/blob-propagator";
import type { BlobReference } from "@blobscan/blob-storage-manager";
import { fixtures, omitDBTimestampFields } from "@blobscan/test";

import { appRouter } from "../src/app-router";
import type { HandleReorgedSlotsInput } from "../src/routers/indexer/handleReorgedSlots";
import { calculateBlobGasPrice } from "../src/routers/indexer/indexData.utils";
import { createTestContext, unauthorizedRPCCallTest } from "./helpers";
import { INPUT_WITH_DUPLICATED_BLOBS, INPUT } from "./indexer.test.fixtures";

describe("Indexer router", async () => {
  let nonAuthorizedCaller: ReturnType<typeof appRouter.createCaller>;
  let authorizedCaller: ReturnType<typeof appRouter.createCaller>;
  let authorizedContext: Awaited<ReturnType<typeof createTestContext>>;

  beforeAll(async () => {
    const ctx = await createTestContext();

    authorizedContext = await createTestContext({ withAuth: true });

    nonAuthorizedCaller = appRouter.createCaller(ctx);
    authorizedCaller = appRouter.createCaller(authorizedContext);
  });

  afterAll(async () => {
    vi.clearAllMocks();
    vi.unstubAllEnvs();
  });

  describe("indexData", () => {
    describe("when authorized", () => {
      beforeEach(async () => {
        await authorizedCaller.indexer.indexData(INPUT);
      });

      it("should index a block correctly", async () => {
        const indexedBlock = await authorizedContext.prisma.block
          .findUnique({
            where: {
              hash: INPUT.block.hash,
            },
          })
          .then((res) => (res ? omitDBTimestampFields(res) : res));

        const {
          blobGasPrice,
          blobAsCalldataGasUsed: _,
          ...remainingParams
        } = indexedBlock ?? {};
        // const expectedBlobAsCalldataGasUsed = INPUT.blobs.reduce(
        //   (acc, b) => acc + getEIP2028CalldataGas(b.data),
        //   0
        // );
        const expectedBlobGasPrice = calculateBlobGasPrice(
          BigInt(INPUT.block.excessBlobGas)
        );

        // TODO: Fix this test
        // expect(blobAsCalldataGasUsed).toBe(expectedBlobAsCalldataGasUsed);
        expect(blobGasPrice?.toString(), "Blob gas price mismatch").toBe(
          expectedBlobGasPrice.toString()
        );
        expect(remainingParams).toMatchInlineSnapshot(`
          {
            "blobGasUsed": "10000",
            "excessBlobGas": "5000",
            "hash": "blockHash2010",
            "number": 2010,
            "slot": 130,
            "timestamp": 2023-09-01T13:50:21.000Z,
          }
        `);
      });

      it("should index transactions correctly", async () => {
        const indexedTxs = await authorizedContext.prisma.transaction
          .findMany({
            where: {
              blockHash: INPUT.block.hash,
            },
            orderBy: {
              hash: "asc",
            },
          })
          .then((r) => r.map(omitDBTimestampFields));
        // const expectedBlobAsCalldataGasUsed = INPUT.transactions.map((tx) =>
        //   INPUT.blobs
        //     .filter((b) => b.txHash === tx.hash)
        //     .reduce((acc, b) => acc + getEIP2028CalldataGas(b.data), 0)
        // );
        // const blobAsCalldataGasUsed = indexedTxs.map(
        //   (tx) => tx.blobAsCalldataGasUsed
        // );
        const remainingParams = indexedTxs.map(
          ({ blobAsCalldataGasUsed: _, ...remainingParams }) => remainingParams
        );

        // TODO: Fix this test
        // expect(
        //   blobAsCalldataGasUsed,
        //   "Transactions' blob as calldata gas used mismatch"
        // ).toEqual(expectedBlobAsCalldataGasUsed);
        expect(remainingParams).toMatchInlineSnapshot(`
          [
            {
              "blockHash": "blockHash2010",
              "fromId": "address7",
              "gasPrice": "3000000",
              "hash": "txHash1000",
              "maxFeePerBlobGas": "20000",
              "rollup": null,
              "toId": "address2",
            },
            {
              "blockHash": "blockHash2010",
              "fromId": "address9",
              "gasPrice": "10000",
              "hash": "txHash999",
              "maxFeePerBlobGas": "1800",
              "rollup": null,
              "toId": "address10",
            },
          ]
        `);
      });

      describe("when indexing blobs", () => {
        it("should index them correctly", async () => {
          const txHashes = INPUT.transactions.map((tx) => tx.hash);
          const indexedBlobs = (
            await authorizedContext.prisma.blobsOnTransactions.findMany({
              where: {
                txHash: {
                  in: txHashes,
                },
              },
            })
          ).sort((a, b) => a.blobHash.localeCompare(b.blobHash));

          expect(indexedBlobs).toMatchInlineSnapshot(`
            [
              {
                "blobHash": "blobHash1000",
                "index": 0,
                "txHash": "txHash1000",
              },
              {
                "blobHash": "blobHash1001",
                "index": 1,
                "txHash": "txHash1000",
              },
              {
                "blobHash": "blobHash999",
                "index": 0,
                "txHash": "txHash999",
              },
            ]
          `);
        });

        describe("when indexing blob data", () => {
          const blobVersionedHashes = INPUT.blobs.map((b) => b.versionedHash);

          describe("when blob propagator is disabled", () => {
            it("should store it on the db correctly", async () => {
              const dbBlobData = await authorizedContext.prisma.blobData
                .findMany({
                  where: {
                    id: {
                      in: blobVersionedHashes,
                    },
                  },
                })
                .then((res) =>
                  res
                    .sort((a, b) => a.id.localeCompare(b.id))
                    .map((b) => `0x${b.data.toString("hex")}`)
                );

              expect(dbBlobData).toMatchInlineSnapshot(`
              [
                "0x34567890abcdef1234567890abcdef",
                "0x34567890abcdef1234567890abcdef1234567890abcdef",
                "0x1234abcdeff123456789ab",
              ]
            `);
            });

            it("should store it on google storage correctly", async () => {
              const blobDataStorageRefs =
                await authorizedContext.prisma.blobDataStorageReference.findMany(
                  {
                    where: {
                      AND: {
                        blobHash: {
                          in: blobVersionedHashes,
                        },
                        blobStorage: "GOOGLE",
                      },
                    },
                  }
                );
              const blobRefs = blobDataStorageRefs.map<BlobReference>(
                (ref) => ({
                  reference: ref.dataReference,
                  storage: "GOOGLE",
                })
              );
              const gcsBlobData = await Promise.all(
                blobRefs.map((ref) =>
                  authorizedContext.blobStorageManager.getBlob(ref)
                )
              ).then((res) =>
                res.sort((a, b) => (a && b ? a.data.localeCompare(b.data) : 0))
              );

              expect(gcsBlobData).toMatchInlineSnapshot(`
                [
                  {
                    "data": "0x1234abcdeff123456789ab",
                    "storage": "GOOGLE",
                  },
                  {
                    "data": "0x34567890abcdef1234567890abcdef",
                    "storage": "GOOGLE",
                  },
                  {
                    "data": "0x34567890abcdef1234567890abcdef1234567890abcdef",
                    "storage": "GOOGLE",
                  },
                ]
              `);
            });

            it("should create blob storage references correctly", async () => {
              const indexedBlobHashes = INPUT.blobs.map(
                (blob) => blob.versionedHash
              );
              const blobStorageRefs =
                await authorizedContext.prisma.blobDataStorageReference.findMany(
                  {
                    orderBy: {
                      blobHash: "asc",
                    },
                    where: {
                      blobHash: {
                        in: indexedBlobHashes,
                      },
                    },
                  }
                );

              expect(blobStorageRefs).toMatchInlineSnapshot(`
                [
                  {
                    "blobHash": "blobHash1000",
                    "blobStorage": "GOOGLE",
                    "dataReference": "70118930558/ob/Ha/sh/obHash1000.txt",
                  },
                  {
                    "blobHash": "blobHash1000",
                    "blobStorage": "POSTGRES",
                    "dataReference": "blobHash1000",
                  },
                  {
                    "blobHash": "blobHash1001",
                    "blobStorage": "GOOGLE",
                    "dataReference": "70118930558/ob/Ha/sh/obHash1001.txt",
                  },
                  {
                    "blobHash": "blobHash1001",
                    "blobStorage": "POSTGRES",
                    "dataReference": "blobHash1001",
                  },
                  {
                    "blobHash": "blobHash999",
                    "blobStorage": "GOOGLE",
                    "dataReference": "70118930558/ob/Ha/sh/obHash999.txt",
                  },
                  {
                    "blobHash": "blobHash999",
                    "blobStorage": "POSTGRES",
                    "dataReference": "blobHash999",
                  },
                ]
              `);
            });
          });
        });

        describe("when blob propagator is enabled", () => {
          let ctxWithBlobPropagator: Awaited<
            ReturnType<typeof createTestContext>
          >;
          let callerWithBlobPropagator: ReturnType<
            typeof appRouter.createCaller
          >;
          let blobPropagatorSpy: SpyInstance<
            [blobs: PropagatorBlob[]],
            Promise<void>
          >;

          beforeAll(async () => {
            ctxWithBlobPropagator = await createTestContext({
              withAuth: true,
              withBlobPropagator: true,
            });

            callerWithBlobPropagator = appRouter.createCaller(
              ctxWithBlobPropagator
            );

            blobPropagatorSpy = vi.spyOn(
              // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
              ctxWithBlobPropagator.blobPropagator!,
              "propagateBlobs"
            );
          });

          afterEach(async () => {
            const blobPropagator = ctxWithBlobPropagator.blobPropagator;

            if (blobPropagator) {
              await blobPropagator.close();
            }
          });

          it("should call blob propagator", async () => {
            await callerWithBlobPropagator.indexer.indexData(
              INPUT_WITH_DUPLICATED_BLOBS
            );

            expect(blobPropagatorSpy).toHaveBeenCalledOnce();
          });
        });
      });

      it("should create valid blob on transaction entities", async () => {
        const indexedBlobsOnTransactions = (
          await authorizedContext.prisma.blobsOnTransactions.findMany({
            where: {
              txHash: {
                in: INPUT.transactions.map((tx) => tx.hash),
              },
            },
          })
        ).sort((a, b) => a.blobHash.localeCompare(b.blobHash));

        expect(indexedBlobsOnTransactions).toMatchInlineSnapshot(`
          [
            {
              "blobHash": "blobHash1000",
              "index": 0,
              "txHash": "txHash1000",
            },
            {
              "blobHash": "blobHash1001",
              "index": 1,
              "txHash": "txHash1000",
            },
            {
              "blobHash": "blobHash999",
              "index": 0,
              "txHash": "txHash999",
            },
          ]
        `);
      });

      it("should indexed addresses correctly", async () => {
        // Remove duplicates
        const addressesSet = new Set(
          INPUT.transactions.flatMap((tx) => [tx.from, tx.to])
        );
        const indexedAddresses = await authorizedContext.prisma.address
          .findMany({
            where: {
              address: {
                in: Array.from(addressesSet),
              },
            },
          })
          .then((r) =>
            r
              .map(omitDBTimestampFields)
              // Prisma doesn't guarantee the order of the results
              .sort((a, b) => a.address.localeCompare(b.address))
          );

        expect(indexedAddresses).toMatchInlineSnapshot(`
          [
            {
              "address": "address10",
              "firstBlockNumberAsReceiver": 2010,
              "firstBlockNumberAsSender": null,
            },
            {
              "address": "address2",
              "firstBlockNumberAsReceiver": 1001,
              "firstBlockNumberAsSender": 1002,
            },
            {
              "address": "address7",
              "firstBlockNumberAsReceiver": null,
              "firstBlockNumberAsSender": 2010,
            },
            {
              "address": "address9",
              "firstBlockNumberAsReceiver": null,
              "firstBlockNumberAsSender": 2010,
            },
          ]
        `);
      });

      it("should be idempotent", async () => {
        // Index the same block again
        await expect(
          authorizedCaller.indexer.indexData(INPUT)
        ).resolves.toBeUndefined();
      });
    });

    it("should fail when calling procedure without auth", async () => {
      await expect(
        nonAuthorizedCaller.indexer.indexData(INPUT)
      ).rejects.toThrow(new TRPCError({ code: "UNAUTHORIZED" }));
    });
  });

  describe("handleReorgedSlots", () => {
    describe("when authorized", () => {
      const input: HandleReorgedSlotsInput = {
        reorgedSlots: [106, 107, 108],
      };

      it("should mark the transactions contained in the blocks with a slot greater than the new head slot as reorged", async () => {
        const prevTransactionForks =
          await authorizedContext.prisma.transactionFork.findMany();
        const expectedTransactionForks = fixtures.txs
          .filter(({ blockHash }) => {
            const block = fixtures.blocks.find(
              ({ hash }) => hash === blockHash
            );

            return block?.slot && input.reorgedSlots.includes(block.slot);
          })
          .map(({ hash, blockHash }) => ({
            hash,
            blockHash,
          }));

        await authorizedCaller.indexer.handleReorgedSlots(input);

        const transactionForks = await authorizedContext.prisma.transactionFork
          .findMany()
          .then((txForks) =>
            txForks
              .map((txFork) => omitDBTimestampFields(txFork))
              .filter(
                (txFork) =>
                  !prevTransactionForks.find(
                    (prevTxFork) => prevTxFork.hash === txFork.hash
                  )
              )
          );

        expect(transactionForks).toEqual(expectedTransactionForks);
      });

      it("should return the number of updated slots", async () => {
        const result = await authorizedCaller.indexer.handleReorgedSlots(input);

        expect(result).toEqual({
          totalUpdatedSlots: input.reorgedSlots.length,
        });
      });

      it("should ignore non-existent slots and mark the ones that exist as reorged", async () => {
        const reorgedSlots = [106, 107, 99999];
        const prevTransactionForks =
          await authorizedContext.prisma.transactionFork.findMany();
        const expectedTransactionForks = fixtures.txs
          .filter(({ blockHash }) => {
            const block = fixtures.blocks.find(
              ({ hash }) => hash === blockHash
            );

            return block?.slot && reorgedSlots.includes(block.slot);
          })
          .map(({ hash, blockHash }) => ({
            hash,
            blockHash,
          }));

        const result = await authorizedCaller.indexer.handleReorgedSlots({
          reorgedSlots: reorgedSlots as [number, ...number[]],
        });

        const transactionForks = await authorizedContext.prisma.transactionFork
          .findMany()
          .then((txForks) =>
            txForks
              .map((txFork) => omitDBTimestampFields(txFork))
              .filter(
                (txFork) =>
                  !prevTransactionForks.find(
                    (prevTxFork) => prevTxFork.hash === txFork.hash
                  )
              )
          );

        expect(transactionForks, "Fork transactions mismatch").toEqual(
          expectedTransactionForks
        );
        expect(
          result.totalUpdatedSlots,
          "Total updated slots mismatch"
        ).toEqual(2);
      });
    });

    it("should not mark any of the provided slots as reorged if all of them are non-existent", async () => {
      const reorgedSlots = [99999, 99998, 99997];
      const prevTransactionForks =
        await authorizedContext.prisma.transactionFork.findMany();

      const result = await authorizedCaller.indexer.handleReorgedSlots({
        reorgedSlots: reorgedSlots as [number, ...number[]],
      });

      const transactionForks = await authorizedContext.prisma.transactionFork
        .findMany()
        .then((txForks) =>
          txForks
            .map((txFork) => omitDBTimestampFields(txFork))
            .filter(
              (txFork) =>
                !prevTransactionForks.find(
                  (prevTxFork) => prevTxFork.hash === txFork.hash
                )
            )
        );

      expect(transactionForks, " Fork transactions mismatch").toEqual([]);
      expect(result.totalUpdatedSlots, "Total updated slots mismatch").toEqual(
        0
      );
    });

    unauthorizedRPCCallTest(() =>
      nonAuthorizedCaller.indexer.handleReorgedSlots({ reorgedSlots: [1000] })
    );
  });
});
