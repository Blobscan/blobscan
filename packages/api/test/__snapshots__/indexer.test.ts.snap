// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Indexer route > indexData > should index new data if auth 1`] = `
{
  "blobAsCalldataGasUsed": 608,
  "blobGasPrice": 1n,
  "blobGasUsed": 10000,
  "excessBlobGas": 5000,
  "hash": "blockHash003",
  "number": 1003,
  "slot": 103,
  "timestamp": 2021-08-29T17:12:54.000Z,
  "transactions": [
    {
      "blobs": [
        {
          "blob": {
            "size": 15,
          },
          "blobHash": "blobHash009",
          "index": 0,
        },
      ],
      "fromId": "address4",
      "hash": "txHash009",
      "toId": "address8",
    },
    {
      "blobs": [
        {
          "blob": {
            "size": 23,
          },
          "blobHash": "blobHash010",
          "index": 0,
        },
      ],
      "fromId": "address7",
      "hash": "txHash010",
      "toId": "address2",
    },
  ],
}
`;

exports[`Indexer route > indexData > should index new data if auth 2`] = `
[
  {
    "address": "address1",
  },
  {
    "address": "address2",
  },
  {
    "address": "address3",
  },
  {
    "address": "address4",
  },
  {
    "address": "address5",
  },
  {
    "address": "address6",
  },
  {
    "address": "address7",
  },
  {
    "address": "address8",
  },
]
`;

exports[`Indexer route > indexData > should index new data if auth 3`] = `
[
  {
    "blockNumber": 1002,
    "hash": "0x5be77167b05f39ea8950f11b0da2bdfec6e04055030068b051ac5a43aaf251e9",
  },
  {
    "blockNumber": 1001,
    "hash": "txHash001",
  },
  {
    "blockNumber": 1001,
    "hash": "txHash002",
  },
  {
    "blockNumber": 1001,
    "hash": "txHash003",
  },
  {
    "blockNumber": 1002,
    "hash": "txHash004",
  },
  {
    "blockNumber": 1002,
    "hash": "txHash005",
  },
  {
    "blockNumber": 1003,
    "hash": "txHash009",
  },
  {
    "blockNumber": 1003,
    "hash": "txHash010",
  },
]
`;

exports[`Indexer route > indexData > should index new data if auth 4`] = `
[
  {
    "commitment": "0xb4f67eb0771fbbf1b06b88ce0e23383daf994320508d44dd30dbd507f598c0d9b3da5a152e41a0428375060c3803b983",
    "firstBlockNumber": 1002,
    "size": 1500,
    "versionedHash": "0x010001c79d78a76fb9b4bab3896ee3ea32f3e2607da7801eb1a92da39d6c1368",
  },
  {
    "commitment": "commitment001",
    "firstBlockNumber": 1001,
    "size": 1000,
    "versionedHash": "blobHash001",
  },
  {
    "commitment": "commitment002",
    "firstBlockNumber": 1001,
    "size": 1100,
    "versionedHash": "blobHash002",
  },
  {
    "commitment": "commitment003",
    "firstBlockNumber": 1001,
    "size": 10,
    "versionedHash": "blobHash003",
  },
  {
    "commitment": "commitment004",
    "firstBlockNumber": 1002,
    "size": 1300,
    "versionedHash": "blobHash004",
  },
  {
    "commitment": "commitment005",
    "firstBlockNumber": 1002,
    "size": 1400,
    "versionedHash": "blobHash005",
  },
  {
    "commitment": "commitment009",
    "firstBlockNumber": 1003,
    "size": 15,
    "versionedHash": "blobHash009",
  },
  {
    "commitment": "commitment010",
    "firstBlockNumber": 1003,
    "size": 23,
    "versionedHash": "blobHash010",
  },
]
`;

exports[`Indexer route > indexData > should index new data if auth 5`] = `
[
  {
    "blobHash": "0x010001c79d78a76fb9b4bab3896ee3ea32f3e2607da7801eb1a92da39d6c1368",
    "blobStorage": "SWARM",
    "dataReference": "bzz://some-hash-for-blobHash006",
  },
  {
    "blobHash": "blobHash001",
    "blobStorage": "GOOGLE",
    "dataReference": "7011893058/ob/Ha/sh/obHash001.txt",
  },
  {
    "blobHash": "blobHash001",
    "blobStorage": "SWARM",
    "dataReference": "bzz://some-hash-for-blobHash001",
  },
  {
    "blobHash": "blobHash002",
    "blobStorage": "GOOGLE",
    "dataReference": "7011893058/ob/Ha/sh/obHash002.txt",
  },
  {
    "blobHash": "blobHash003",
    "blobStorage": "SWARM",
    "dataReference": "bzz://some-hash-for-blobHash003",
  },
  {
    "blobHash": "blobHash004",
    "blobStorage": "GOOGLE",
    "dataReference": "7011893058/ob/Ha/sh/obHash004.txt",
  },
  {
    "blobHash": "blobHash004",
    "blobStorage": "POSTGRES",
    "dataReference": "blobHash004",
  },
  {
    "blobHash": "blobHash005",
    "blobStorage": "POSTGRES",
    "dataReference": "blobHash005",
  },
  {
    "blobHash": "blobHash009",
    "blobStorage": "GOOGLE",
    "dataReference": "7011893058/ob/Ha/sh/obHash009.txt",
  },
  {
    "blobHash": "blobHash009",
    "blobStorage": "POSTGRES",
    "dataReference": "blobHash009",
  },
  {
    "blobHash": "blobHash010",
    "blobStorage": "GOOGLE",
    "dataReference": "7011893058/ob/Ha/sh/obHash010.txt",
  },
  {
    "blobHash": "blobHash010",
    "blobStorage": "POSTGRES",
    "dataReference": "blobHash010",
  },
]
`;

exports[`Indexer route > indexData > should index new data if auth 6`] = `
[
  {
    "blobHash": "0x010001c79d78a76fb9b4bab3896ee3ea32f3e2607da7801eb1a92da39d6c1368",
    "index": 0,
    "txHash": "0x5be77167b05f39ea8950f11b0da2bdfec6e04055030068b051ac5a43aaf251e9",
  },
  {
    "blobHash": "blobHash001",
    "index": 0,
    "txHash": "txHash001",
  },
  {
    "blobHash": "blobHash002",
    "index": 0,
    "txHash": "txHash002",
  },
  {
    "blobHash": "blobHash003",
    "index": 0,
    "txHash": "txHash003",
  },
  {
    "blobHash": "blobHash004",
    "index": 0,
    "txHash": "txHash004",
  },
  {
    "blobHash": "blobHash005",
    "index": 0,
    "txHash": "txHash005",
  },
  {
    "blobHash": "blobHash009",
    "index": 0,
    "txHash": "txHash009",
  },
  {
    "blobHash": "blobHash010",
    "index": 0,
    "txHash": "txHash010",
  },
]
`;
